### PR TITLE
MSBuild warnings as errors

### PR DESCRIPTION
**Pull request by jaredpar**
Use the /warnaserror flag that is a part of MSBuild 15.0.

### CODE REVIEW

**Comment by jaredpar** [comment]
CC @dotnet/roslyn-infrastructure

**Comment by jaredpar** [comment]
CC @jeffkl to see his feature in action.

File: build/scripts/build-utils.ps1 [hunk 1]
```diff
@@ -13,7 +13,7 @@ $ErrorActionPreference="Stop"
 # Original sample came from: http://jameskovacs.com/2010/02/25/the-exec-problem/
 function Exec([scriptblock]$cmd, [string]$errorMessage = "Error executing command: " + $cmd) { 
     $output = & $cmd 
-    if (-not $?) {
+    if ((-not $?) -or ($lastexitcode -ne 0)) {
```
Comment (jasonmalinowski) [inline build/scripts/build-utils.ps1]: This only reinforces my belief that powershell is the best terrible scripting language we have.

File: src/Tools/MicroBuild/cibuild.ps1 [hunk 2]
```diff
@@ -22,7 +22,11 @@ try {
     Exec { & $nuget locals all -clear }
 
     $msbuild = Ensure-MSBuild
-    & $msbuild /nodereuse:false /p:Configuration=Release /p:SkipTest=true Build.proj 
+
+    # The /nowarn exception can be removed once we fix the following:
+    #  - 3277: https://github.com/dotnet/roslyn/issues/17325
+    #  - 4011: https://github.com/dotnet/roslyn/issues/18151
+    & $msbuild /nodereuse:false /p:Configuration=Release /p:SkipTest=true Build.proj /warnaserror /nowarn:3277,4011
```
Comment (jasonmalinowski) [inline src/Tools/MicroBuild/cibuild.ps1]: This doesn't break actual microbuild? Can you confirm we're on the Dev15 pool at this point before merging?

Comment (tannergooding) [inline src/Tools/MicroBuild/cibuild.ps1]: MicroBuild doesn't go through cibuild.ps1, IIRC.

Comment (jaredpar) [inline src/Tools/MicroBuild/cibuild.ps1]: @tannergooding is correct.  Microbuild enters us directly at Build.proj.  In Jenkins cibuild.ps1 drives our entry into Build.proj.

**Comment by jeffkl** [comment]
FYI there is a bug with the /warnaserror and multi-proc builds where it some times won't fail the build.  The errors would be logged but MSBuild.exe will have an exit code of zero.  It turned out to be race condition only in optimized, ngen'd MSBuild.  It should only happen if the only logged error occurs as one of the very last messages and the build ends.  I've already fixed the bug but it was reported after we shipped.

https://github.com/Microsoft/msbuild/issues/1791

**Comment by jaredpar** [comment]
>  The errors would be logged but MSBuild.exe will have an exit code of zero. It turned out to be race condition only in optimized, ngen'd MSBuild. 

Good new is we use xcopy msbuild on many of our builds so hopefully won't hit this.

Has this fix been included in a NuGet package yet?  if so I can just consume it dircetly.

**Comment by jeffkl** [comment]
Yes, the latest package `15.2.0-preview-000083-02` should have the fix.

**Comment by jaredpar** [comment]
@jeffkl great.  thanks!

**Comment by jaredpar** [comment]
retest microbuild_prtest please

**Comment by sharwell** [comment]
@jaredpar

:question: Do we not want to set this in the project properties as well (for developers)?

:question: When both rule set files and this flag are used, which has precedence?

**Comment by jaredpar** [comment]
@sharwell 

> Do we not want to set this in the project properties as well (for developers)?

Didn't realize this was also available as a property.  Have to figure out if I can properly opt out #17325 with a local change though.  Also unclear how to set the "all" with `<WarningsAsError>` as the docs are ambiguous.  Will pose a question on MSBuild.  

> When both rule set files and this flag are used, which has precedence?

Will have to let @jeffkl handle that.

**Comment by jeffkl** [comment]
The first iteration of this feature is only available at the command-line or through the object model.  Having an MSBuild property to control warnings is a little tricky because a set of loggers are instantiated for the entire build episode.  The list of warnings to treat as errors is a single map for the whole build and so individual projects can't control warnings.  We also log warnings during project evaluation which would not take into account project-level warnings defined in properties.  If we get enough feedback to improve the feature, we'll do it.

Leave your votes here: https://github.com/Microsoft/msbuild/issues/1886

**Comment by jaredpar** [comment]
@jeffkl gotcha.  Which comment should we vote on?  Basically I want to know where I can click to say "Let me global property this" 😄

**Comment by jeffkl** [comment]
Vote on Microsoft/msbuild#1886

**Pull request merged**